File: calendar/lunar/lunar_events.js
const Astronomy = require('astronomy-engine');
const fs = require('fs');

function findLunarEvents(startDate = '2025-05-26', days = 365) {
  const start = new Astronomy.AstroTime(new Date(startDate + 'T00:00:00Z'));
  const end = start.AddDays(days);

  let events = [];

  let mq = Astronomy.SearchMoonQuarter(start);
  while (mq.time.tt <= end.tt) {
    if (mq.quarter === 0) events.push({ time: mq.time.date, event: 'New Moon' });
    else if (mq.quarter === 2) events.push({ time: mq.time.date, event: 'Full Moon' });
    mq = Astronomy.NextMoonQuarter(mq);
  }

  let aps = Astronomy.SearchLunarApsis(start);
  while (aps.time.tt <= end.tt) {
    if (aps.kind === Astronomy.ApsisKind.Pericenter) {
      events.push({ time: aps.time.date, event: 'Perigee' });
    } else {
      events.push({ time: aps.time.date, event: 'Apogee' });
    }
    aps = Astronomy.NextLunarApsis(aps);
  }

  events.sort((a, b) => a.time - b.time);

  const midpoints = [];
  for (let i = 0; i < events.length - 1; i++) {
    const t0 = events[i].time.getTime();
    const t1 = events[i + 1].time.getTime();
    const mid = new Date((t0 + t1) / 2);
    midpoints.push({ time: mid, event: `Midpoint ${events[i].event}->${events[i+1].event}` });
  }

  return { events, midpoints };
}

if (require.main === module) {
  const { events, midpoints } = findLunarEvents();
  const format = (d) => d.toISOString();
  const eventLines = ['time,event'];
  events.forEach(e => eventLines.push(`${format(e.time)},${e.event}`));
  fs.writeFileSync('lunar_events.csv', eventLines.join('\n'));

  const midLines = ['time,event'];
  midpoints.forEach(m => midLines.push(`${format(m.time)},${m.event}`));
  fs.writeFileSync('lunar_midpoints.csv', midLines.join('\n'));
  console.log('Saved lunar_events.csv and lunar_midpoints.csv');
}

module.exports = { findLunarEvents };

File: calendar/lunar/lunar_events.py
import pandas as pd
from datetime import datetime
from skyfield.api import load
from skyfield import almanac


def find_lunar_events(start_date="2025-05-26", days=365):
    ts = load.timescale()
    eph = load('de421.bsp')
    start = ts.utc(datetime.fromisoformat(start_date))
    end = start + days

    # Moon phases: 0=new, 2=full
    t_phase, y_phase = almanac.find_discrete(start, end, almanac.moon_phases(eph))
    # Moon apsides: 0=perigee, 1=apogee
    t_apsides, y_apsides = almanac.find_discrete(start, end, almanac.moon_apsides(eph))

    events = []
    for t, phase in zip(t_phase, y_phase):
        if phase == almanac.NEW_MOON:
            events.append({"time": t.utc_datetime(), "event": "New Moon"})
        elif phase == almanac.FULL_MOON:
            events.append({"time": t.utc_datetime(), "event": "Full Moon"})

    for t, apsis in zip(t_apsides, y_apsides):
        if apsis == 0:
            events.append({"time": t.utc_datetime(), "event": "Perigee"})
        else:
            events.append({"time": t.utc_datetime(), "event": "Apogee"})

    events.sort(key=lambda x: x["time"])

    midpoints = []
    for i in range(len(events) - 1):
        t0 = events[i]["time"]
        t1 = events[i + 1]["time"]
        mid = t0 + (t1 - t0) / 2
        midpoints.append({"time": mid, "event": f"Midpoint {events[i]['event']} -> {events[i+1]['event']}"})

    df_events = pd.DataFrame(events)
    df_mid = pd.DataFrame(midpoints)
    return df_events, df_mid


if __name__ == "__main__":
    events, mids = find_lunar_events()
    events.to_csv("lunar_events.csv", index=False)
    mids.to_csv("lunar_midpoints.csv", index=False)
    print("Saved lunar_events.csv and lunar_midpoints.csv")
